### DISCUSSION TITLE
How to use NodeJs based language server in Visual Studio as Language Server?

### DISCUSSION THREAD

**Discussion by soroshsabz** [VsixCommunity/Community.VisualStudio.Toolkit#424] [Q&A] [answered]
I want to add ansible DSL (based on yaml) support for Visual Studio 2022 (with extension), I found a language server for ansible DSL in [visual studio code](https://github.com/ansible/ansible-language-server), and I want to port this language server into [my extension](https://github.com/soroshsabz/visualstudio-ansible/issues/4) for Visual Studio 2022.

**Can anybody point to any references or even examples of existing porting?**

I think porting with minimal changing is very useful, because we can update language server based on original language server with every changes.

I request sample in https://github.com/microsoft/VSSDK-Extensibility-Samples/issues/279#issue-1589674661

**Comment by soroshsabz** [comment]
I ask same question in [SO](https://stackoverflow.com/q/75469697/1539100)

**Comment by soroshsabz** [comment]
@madskristensen Did you have any idea, or suggestion about this?

**Comment by CalvinAllen** [comment, marked as answer]
Just start it as an external process during the activation of the Language Server

Start here for an example - https://github.com/TeamCodeStream/codestream/blob/develop/vs/src/CodeStream.VisualStudio.Shared/LanguageServer/LanguageClient.cs#L120

**Reply by CalvinAllen** [reply]
Follow that down and you can see the using of Process here -

https://github.com/TeamCodeStream/codestream/blob/develop/vs/src/CodeStream.VisualStudio.Shared/LanguageServer/LanguageServerClientProcess.cs and then down to https://github.com/TeamCodeStream/codestream/blob/develop/vs/src/CodeStream.VisualStudio.Core/Process/ProcessFactory.cs

We ship an executable version for Release and just the raw JS for Debug.

**Reply by soroshsabz** [reply]
Thanks a lot, for responding, I read your code, and it is very similar to my need.

I have another question from your code, How to inject some dependencies in `LanguageClient`? in your code, you put many dependencies in `LanguageClient` constructor, but I do not know, how to doing something like that? when I try to add some arguments for `LanguageClient` constructor, I got exception. how to resolve this?

thanks

**Reply by soroshsabz** [reply]
@CalvinAllen

**Reply by CalvinAllen** [reply]
@soroshsabz Those are just other "services" that have been flagged with the `Export()` attribute for usage with MEF. For instance, the implementation of `IHttpClientService` here - https://github.com/TeamCodeStream/codestream/blob/develop/vs/src/CodeStream.VisualStudio.Shared/Services/HttpClientService.cs#L14

**Reply by soroshsabz** [reply]
thanks a lot

**Reply by CalvinAllen** [reply]
You're welcome, good luck! :)

**Reply by soroshsabz** [reply]
@CalvinAllen I apologize for taking up your time so often, but I have another problem, I write my code and push to https://github.com/soroshsabz/visualstudio-ansible/tree/features/4-support-ansible-language-server , my problem is, when I run visual studio in debugging mode, `ActivateAsync` method, calls frequently, infinitely. My question is why? and how to correct this problem?

**Reply by CalvinAllen** [reply]
@soroshsabz I'll take a look...not sure off-hand since that method is (supposed) to be called/handled by Visual Studio when you open a file of the declared type

**Reply by soroshsabz** [reply]
@CalvinAllen thanks a lot, My problem is when open `.ansible` file, Visual Studio infinitely activate language server over and over again, and I do not know my mistake to cause this problem

**Reply by CalvinAllen** [reply]
@soroshsabz There may be some exception happening somewhere causing the LSP not to start, so it keeps trying / looping - do you generate any log files or messages anywhere? Might be a good time to add something if not

**Reply by soroshsabz** [reply]
In background task running window (in bottom of visual studio), visual studio says explicitly to `Task Completed Successfully` for Activating Language Client: Ansible Language Extension and sometimes says `Client Initialization Completed`

I do not know where of code need logging, because I debug step by step in project, and exception does not rises, in anywhere.

**Reply by soroshsabz** [reply]
@CalvinAllen Is default logging exist anywhere, to I can see it, and find problem?

thanks a lot

**Reply by CalvinAllen** [reply]
@soroshsabz Not to my knowledge - we just write out stuff to a log file on disk, but you could probably bring in any logging framework and utilize it.
